docs/v1/guide/best-practices.md

```diff
@@ -0,0 +1,600 @@
+presenceData.state = 'Email: user@example.com'
+```
+
+### Validate User Input
+
+If your activity uses user input, validate it to prevent security vulnerabilities.
+
+```typescript
+// Good
+const userInput = document.querySelector('.user-input')?.textContent || ''
+const sanitizedInput = userInput.replace(/[<>]/g, '') // Remove < and > to prevent HTML injection
+
+presenceData.details = sanitizedInput
+
+// Bad
+const userInput = document.querySelector('.user-input')?.textContent || ''
+presenceData.details = userInput
+```
+
+## Complete Example
+
+Here's a complete example of an activity that follows best practices:
+
+### metadata.json
+
+```json
+{
+  "author": {
+    "name": "Your Name",
+    "id": "your_discord_id"
+  },
+  "service": "Example",
+  "description": {
+    "de": "Example ist eine Website, die etwas Cooles macht.",
+    "en": "Example is a website that does something cool.",
+    "fr": "Example est un site web qui fait quelque chose de cool."
+  },
+  "url": "example.com",
+  "version": "1.0.0",
+  "logo": "https://example.com/logo.png",
+  "thumbnail": "https://example.com/thumbnail.png",
+  "color": "#FF0000",
+  "category": "other",
+  "tags": ["example", "best-practices"],
+  "settings": [
+    {
+      "id": "showButtons",
+      "multiLanguage": true,
+      "value": true
+    },
+    {
+      "id": "showTimestamp",
+      "multiLanguage": true,
+      "value": true
+    },
+    {
+      "id": "showDetails",
+      "multiLanguage": true,
+      "value": true
+    }
+  ]
+}
+```
+
+### presence.ts
+
+```typescript
+const presence = new Presence({
+  clientId: 'your_client_id'
+})
+
+// Constants
+const DEFAULT_IMAGE_KEY = 'logo'
+const DEFAULT_TIMESTAMP = Date.now()
+
+// Cache for heavy computations
+const cachedData = null
+const lastComputation = 0
+
+// Helper function to get page information
+function getPageInfo() {
+  const path = document.location.pathname
+  const title = document.title
+
+  let pageType = 'unknown'
+
+  if (path === '/') {
+    pageType = 'homepage'
+  }
+  else if (path.includes('/about')) {
+    pageType = 'about'
+  }
+  else if (path.includes('/contact')) {
+    pageType = 'contact'
+  }
+
+  return { path, title, pageType }
+}
+
+// Helper function to get video information
+function getVideoInfo() {
+  const video = document.querySelector('video')
+
+  if (!video || video.readyState === 0) {
+    return null
+  }
+
+  return {
+    title: document.querySelector('.video-title')?.textContent || 'Unknown video',
+    author: document.querySelector('.video-author')?.textContent || 'Unknown author',
+    isPlaying: !video.paused,
+    currentTime: video.currentTime,
+    duration: video.duration
+  }
+}
+
+presence.on('UpdateData', async () => {
+  // Get settings
+  const showButtons = await presence.getSetting<boolean>('showButtons')
+  const showTimestamp = await presence.getSetting<boolean>('showTimestamp')
+  const showDetails = await presence.getSetting<boolean>('showDetails')
```

Comment:
not best practice

Reply:
getSetting internally isn't async anymore so this has no performance implications I just left it as async for compatibility sake @Bas950 ^

Reply:
No but wrap it in an Promise.all like:

```
const [
    showButtons,
    showTimestamps,
    showCover,
    hidePaused,
    showBrowsing,
    privacyMode,
    useTimeLeft,
    showAsListening,
  ] = await Promise.all([
    presence.getSetting<boolean>('buttons'),
    presence.getSetting<boolean>('timestamps'),
    presence.getSetting<boolean>('cover'),
    presence.getSetting<boolean>('hidePaused'),
    presence.getSetting<boolean>('browsing'),
    presence.getSetting<boolean>('privacy'),
    presence.getSetting<boolean>('useTimeLeft'),
    presence.getSetting<boolean>('showAsListening'),
  ])
  ```